internal/exec/go_getter_utils.go

```diff
@@ -56,77 +58,165 @@ func IsValidScheme(scheme string) bool {
-		usedToken = atmosGitHubToken
+	// 3 types of tokens are supported for now: Github, Bitbucket and GitLab
+	var token, tokenSource string
+	switch host {
+	case "github.com":
 		tokenSource = "ATMOS_GITHUB_TOKEN"
-		u.LogDebug("ATMOS_GITHUB_TOKEN is set\n")
-	} else {
-		// 2. Otherwise, only inject GITHUB_TOKEN if cfg.Settings.InjectGithubToken == true
-		if d.AtmosConfig.Settings.InjectGithubToken && gitHubToken != "" {
-			usedToken = gitHubToken
+		token = os.Getenv(tokenSource)
+		if token == "" && d.AtmosConfig.Settings.InjectGithubToken {
 			tokenSource = "GITHUB_TOKEN"
-			u.LogTrace("InjectGithubToken=true and GITHUB_TOKEN is set, using it\n")
-		} else {
-			u.LogTrace("No ATMOS_GITHUB_TOKEN or GITHUB_TOKEN found\n")
+			token = os.Getenv(tokenSource)
+		}
+	case "bitbucket.org":
+		tokenSource = "ATMOS_BITBUCKET_TOKEN"
+		token = os.Getenv(tokenSource)
+		if token == "" {
+			tokenSource = "BITBUCKET_TOKEN"
+			token = os.Getenv(tokenSource)
+		}
+	case "gitlab.com":
+		tokenSource = "ATMOS_GITLAB_TOKEN"
+		token = os.Getenv(tokenSource)
+		if token == "" {
+			tokenSource = "GITLAB_TOKEN"
+			token = os.Getenv(tokenSource)
 		}
 	}
 
-	if usedToken != "" {
-		user := parsedURL.User.Username()
-		pass, _ := parsedURL.User.Password()
-		if user == "" && pass == "" {
-			u.LogDebug(fmt.Sprintf("Injecting token from %s for %s\n", tokenSource, src))
-			parsedURL.User = url.UserPassword("x-access-token", usedToken)
+	// Note that Bitbucket uses 2 tokens (username and app password) for authentication.
+	if token != "" {
+		// Inject token only if no credentials are already provided.
+		if parsedURL.User == nil || parsedURL.User.Username() == "" {
+			l.Debug("Injecting token", "token_source", tokenSource, "url", src)
+			var defaultUsername string
+			switch host {
+			case "github.com":
+				defaultUsername = "x-access-token"
+			case "gitlab.com":
+				defaultUsername = "oauth2"
+			case "bitbucket.org":
+				defaultUsername = os.Getenv("BITBUCKET_USERNAME")
+				if defaultUsername == "" {
+					defaultUsername = "x-token-auth"
+				}
+				l.Debug("Using Bitbucket username", "username", defaultUsername)
+			default:
+				defaultUsername = "x-access-token"
+			}
+			parsedURL.User = url.UserPassword(defaultUsername, token)
 		} else {
-			u.LogDebug("Credentials found, skipping token injection\n")
+			l.Debug("Skipping token injection", "reason", "credentials already provided")
+		}
+	}
+
+	// Normalize d.source for Windows path separators.
+	normalizedSource := filepath.ToSlash(d.source)
+	// If d.source is provided (non‑empty), use it for subdir checking;
+	// otherwise, skip appending '//.' (so the user-defined subdir isn’t mistakenly processed).
+	if normalizedSource != "" && !strings.Contains(normalizedSource, "//") {
+		parts := strings.SplitN(parsedURL.Path, "/", 4)
+		if strings.HasSuffix(parsedURL.Path, ".git") || len(parts) == 3 {
+			l.Debug("Detected top-level repo with no subdir: appending '//.'", "url", src)
```

Comment:
I will this leak the injected secret?

Reply:
Please add a test where the vendor configuration we set a basic auth, like “supersecret” then enable debug output and expect this string is nowhere in the output.